api/odl.py

```diff
@@ -0,0 +1,598 @@
+        READY_STATUS,
+        ACTIVE_STATUS,
+        REVOKED_STATUS,
+        RETURNED_STATUS,
+        CANCELLED_STATUS,
+        EXPIRED_STATUS,
+    ]
+
+    def __init__(self, _db, collection):
+        if collection.protocol != self.NAME:
+            raise ValueError(
+                "Collection protocol is %s, but passed into ODLWithConsolidatedCopiesAPI!" %
+                collection.protocol
+            )
+        self.collection_id = collection.id
+        self.data_source_name = collection.external_integration.setting(Collection.DATA_SOURCE_NAME_SETTING).value
+        # Create the data source if it doesn't exist yet.
+        DataSource.lookup(_db, self.data_source_name, autocreate=True)
+
+        self.username = collection.external_integration.username
+        self.password = collection.external_integration.password
+        self.consolidated_loan_url = collection.external_integration.setting(self.CONSOLIDATED_LOAN_URL_KEY).value
+        self.default_loan_period = collection.external_integration.setting(BaseCirculationAPI.DEFAULT_LOAN_PERIOD).int_value or 21
+
+    def internal_format(self, delivery_mechanism):
+        """Each consolidated copy is only available in one format, so we don't need
+        a mapping to internal formats.
+        """
+        return delivery_mechanism
+
+    def collection(self, _db):
+        return get_one(_db, Collection, id=self.collection_id)
+
+    def _get(self, url, headers=None):
+        """Make a normal HTTP request, but include an authentication
+        header with the credentials for the collection.
+        """
+
+        username = self.username
+        password = self.password
+        headers = dict(headers or {})
+        auth_header = "Basic %s" % base64.b64encode("%s:%s" % (username, password))
+        headers['Authorization'] = auth_header
+
+        return HTTP.get_with_timeout(url, headers=headers)
+
+    def _url_for(self, *args, **kwargs):
+        """Wrapper around flask's url_for to be overridden for tests.
+        """
+        return url_for(*args, **kwargs)
+
+    def get_license_status_document(self, pool, patron, loan=None):
+        """Get the License Status Document for a loan. This will create the 
+        remote loan if one doesn't exist yet.
+        """
+        _db = Session.object_session(pool)
+
+        id = pool.identifier.identifier
+        if loan:
+            url = loan.external_identifier
+        else:
+            checkout_id = str(uuid.uuid1())
+            expires = datetime.datetime.utcnow() + datetime.timedelta(days=self.default_loan_period)
+            # The patron UUID is generated randomly on each loan, so the distributor
+            # doesn't know when multiple loans come from the same patron.
+            patron_id = str(uuid.uuid1())
+            notification_url = self._url_for(
+                "odl_notify",
+                library_short_name=patron.library.short_name,
+                loan_external_identifier=checkout_id,
```

Comment:
This will no longer work due to my change to make the external identifier the URL instead of the UUID. I'll need to come up with something else for the notify route.

Reply:
You get to choose this URL, right? Can you use the loan's internal database ID?

Reply:
I have to send the URL as part of the request to create the loan, so the loan isn't in the db at this point. I could create it and then delete it after if it didn't work.

Reply:
I think that's better than creating another place to store the UUID, because once you have two extra pieces of information associated with each loan, we need to talk about reusing the `resources` table or something. 

Maybe create the Loan object with as little information as possible just to get the ID, and then fill in the rest later.

But we'll need to make sure that everything works in case someone gets an extra Loan with no associated LicensePool.

Reply:
I should say I'm not opposed to creating a new database table for keeping track of information that dies with the loan. Reusing `resources` is probably not a good idea since that's full of bibliographic information that never dies.

Reply:
I changed it to create the loan object with just the patron and pool and use its id, then delete the loan if the checkout fails.